### DISCUSSION TITLE
Middleware with parameters to require subject/ldap-group

### DISCUSSION THREAD

**Discussion by bram-nijenkamp** [authelia/authelia#6946] [Ideas]
### Description

Been using your porject for a while and love it but something I could not find anywhere is how to require groups or rather the subject in a middleware configuration as opposed to authelia's native access control.

### Use Case

At the moment I very much enjoy the traefik-docker and authelia-middleware method and as such you might imagine that I don't want to add an access control policy or domain routing in to places. I have some services that have no authentication mechanisms at all but I don't want all users in my ldap configuration to go there.

### Details

To me it would be kind of amazing if in the traefik middleware registration I would be able to add a GET/POST parameter in which I could require a a user to be in a ldap group (I use lldap).

### Pre-Submission Checklist

- [X] I agree to follow the [Code of Conduct](http://www.authelia.com/code-of-conduct)

- [X] I have checked for related issues and checked the documentation

**Comment by nightah** [comment]
Keen to understand your thinking here wouldn't a separate middleware for each service be much more overhead to manage rather than ACLs?

Maybe I'm misunderstanding but I don't see why an ACL with a subject limited to groups wouldn't satisfy this requirement?

**Comment by bram-nijenkamp** [comment]
Well first maybe the context I am a mere self-hosted addict. As such I don’t expect that many groups. But it is also the place where you have to define them. What I find so amazing about working with authelia en docker-traefik is the ability to add middleware domain and everything from a single in my case docker compose file. On these I already have middleware for authelia set. Let’s say I host around 35 different services of which i would want to keep 30 to myself but set open some others for say friends or otherwise. The only thing I would need to do is inline the group dependency all together or create one more default middleware and assign it to those containers. If I were to do it with the ACLs I’d have to keep track of the all domains (and should they grow in groups or public domains) I have to double manage those in different configs. This is also the reason why I went away from nginx and caddy. They were all great but the best part is with traefik once again no additional configs aside from the base.

I’d be happy to answer further questions.

**Comment by james-d-elliott** [comment]
I'd planned on commenting on this. I have a few apprehensions, and unfortunately I don't feel like they are small hurdles. The specific ones I can recall right now are:

1. Disconnect between ACLs and the Middleware:
    1. How would someone manage the disconnect between ACLs and the parameters? 
    2. How are we going to avoid users not understanding how it works? 
    3. Couldn't users theoretically specify ACLs that match the request that also include a username/group criteria and wouldn't this lead to fairly confusing outcomes when specified at the same time?
2. How would we implement this in a way that will be compatible with the Traefik Forward Auth middleware but also that will prevent well crafted request from adjusting the auth requirements on the Envoy ExtAuthz filter considering it just appends the entire path to the request? While we're adding implementation specific configurations I think this is quite a divergent feature.

**Comment by bram-nijenkamp** [comment]
1.1. I think the ACL should always have priority over what comes in through the middleware.
1.2. We can just side document it, it's not like this should be the way or anything but for those that need/want to do it this way I don't find it weird. If you want to promote ACLs then by all means keep doing that!
1.3 If the request is in the ACL always use the ACL you can think of merging them but that just adds difficulty...
2. To be fair I have never used envoy nor was familiar with it. We could imagine making this new behavior disabled by default and let people enable it only when they want to and know that is is compatible with their network infrastructure.

**Comment by nightah** [comment]
With the release of 4.38, the team has been discussing this feature request again.

We're relatively confident that safely implementing this feature across the different applications of Authelia is non-trivial and doesn't provide an overly large benefit to a user/administrator.

As such we are going to convert this to a discussion so users can vote to see if our views are aligned with those of our userbase, if someone wants to pick up and attempt to design this we welcome you to do so but it would be worthwhile having some open dialogue with the team to find a potential way forward.

**Comment by glime-ai** [comment]
Please note this response was created using generative AI leveraging previously available information from the repository. This information may be helpful however it is our strong recommendation that you carefully consider the information before blindly applying it.

---

Based on the information provided and the context from the sources, it seems you're looking for a way to integrate LDAP group requirements directly into the Traefik middleware configuration when using Authelia, without having to rely solely on Authelia's native access control lists (ACLs) for domain routing and access control policies. This approach aims to simplify the management of access control by embedding group requirements directly in the middleware configuration, potentially in a single file like a Docker Compose configuration, which aligns with your use case of managing multiple services with varying access requirements.

However, from the discussions and issues reviewed, it appears that Authelia does not currently support specifying LDAP group requirements directly in the Traefik middleware configuration through GET/POST parameters or similar means. The discussions highlight a desire for such a feature but also outline several challenges and considerations that would need to be addressed to implement it, such as managing the potential disconnect between ACLs and middleware parameters, ensuring compatibility with different proxy configurations, and preventing security vulnerabilities.

The current recommended approach for controlling access based on LDAP groups with Authelia and Traefik involves using Authelia's ACLs to define policies that include group-based subjects. These ACLs are then enforced by Authelia during the authentication and authorization process, as configured in the middleware. While this method requires managing ACLs separately from the middleware configuration, it provides a robust and flexible way to control access based on user groups and other criteria.

For your specific use case, you might consider organizing your ACLs to minimize duplication and simplify management, possibly by grouping services with similar access requirements under common domains or subdomains and using wildcard or regex patterns in your ACL definitions. Additionally, keeping an eye on Authelia's updates and community discussions could inform you of any new features or approaches that might better align with your desired workflow.

This comment was generated by [Glime](https://glimelab.ai/?utm_source=6fsHkcyufjQLaB&utm_medium=github).

**Comment by vbouchaud** [comment]
Hi there, this is something I thought of before and while I never took the time to implement it, I'll just explain the concept here:

You could develop a custom ForwardAuth middleware and use it in chain with Authelia. The flow would be:

1. Authelia handles authentication
2. Your custom middleware handles authorization

The Traefik configuration would look like:
```yaml
middlewares:
  auth-chain:
    chain:
      middlewares:
        - authelia-auth    # Authentication via Authelia
        - custom-authz     # Your authorization middleware
```

However, you would need to use something else to inject specific authorization rules so you don't have to build a specific authz mw for each route. I suppose you could use the Headers middleware to do that through custom headers. Your middleware could then:

- Read the authentication headers provided by Authelia (groups, user...)
- Compare this information with the rules defined in the custom headers

A basic example:
```yaml
middlewares:
  my-app-headers:
    headers:
      customRequestHeaders:
        X-Required-Groups: "admin,users"
  
routers:
  my-app:
    rule: Host(`app.example.com`)
    middlewares:
      - auth-chain
      - my-app-headers
```

Well, it's certainly more complex and at least a bit unconventional but it would offer several benefits:

- No custom code in authelia
- Authorization configuration directly in IngressRoute/docker-compose files as manifest, so no need to redeploy/restart authelia
- Increased authz portability for Kubernetes deployments (via Helm for instance)
- Clear separation between authentication and authorization. It could also be used after other authentication app.

I'm not sure about the security implications of using headers to send configuration (header size limit, etc.). I think you would also need clean incoming headers to avoid configuration injection before adding you own. This is not ideal but it could be of help.

The other, safer and better way would probably be to have some custom code read annotations on middlewares/ingressroutes to dynamically rebuild authelia configuration and reload it or, if I keep the idea of using another forwardauth middleware, one could even implement an operator with some CRDs to handle authz configuration, linked to the middleware definition with annotations.
